consensus/src/bullshark.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2021, Facebook, Inc. and its affiliates
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::{
+    consensus::{ConsensusProtocol, ConsensusState, Dag},
+    utils, ConsensusOutput,
+};
+use config::{Committee, SharedCommittee, Stake};
+use crypto::{
+    traits::{EncodeDecodeBase64, VerifyingKey},
+    Hash,
+};
+use std::{collections::HashMap, sync::Arc};
+use tracing::debug;
+use types::{Certificate, CertificateDigest, ConsensusStore, Round, SequenceNumber, StoreResult};
+
+#[cfg(test)]
+#[path = "tests/bullshark_tests.rs"]
+pub mod bullshark_tests;
+
+pub struct Bullshark<PublicKey: VerifyingKey> {
+    /// The committee information.
+    pub committee: SharedCommittee<PublicKey>,
+    /// Persistent storage to safe ensure crash-recovery.
+    pub store: Arc<ConsensusStore<PublicKey>>,
+    /// The depth of the garbage collector.
+    pub gc_depth: Round,
+}
+
+impl<PublicKey: VerifyingKey> ConsensusProtocol<PublicKey> for Bullshark<PublicKey> {
+    fn process_certificate(
+        &mut self,
+        state: &mut ConsensusState<PublicKey>,
+        consensus_index: SequenceNumber,
+        certificate: Certificate<PublicKey>,
+    ) -> StoreResult<Vec<ConsensusOutput<PublicKey>>> {
+        debug!("Processing {:?}", certificate);
+        let round = certificate.round();
+        let mut consensus_index = consensus_index;
+
+        // Add the new certificate to the local storage.
+        state
+            .dag
+            .entry(round)
+            .or_insert_with(HashMap::new)
+            .insert(certificate.origin(), (certificate.digest(), certificate));
+
+        // Try to order the dag to commit. Start from the highest round for which we have at least
+        // 2f+1 certificates. This is because we need them to reveal the common coin.
```

Comment:
For BabyShark even f+1 certificates are enough, in case we want to try reduce latency. 

There is also the 3f+1 (or max allowed) path to reduce latency as discussed. Should I add a conversation?